### PR TITLE
feat: add ga events forwarder plugin

### PR DESCRIPTION
**Pull request by kevinpagtakhan**
### Summary

Adds `@amplitude/plugin-ga-events-forwarder-browser` package

Example usage:

```html
 <script>
   amplitude.add(gaEventsForwarder.plugin());
   amplitude.init(<API_KEY>);
 </script>
 <script async src="https://www.googletagmanager.com/gtag/js?id=G-XXXXXXXXXX"></script>
 <script>
   window.dataLayer = window.dataLayer || [];
   function gtag(){
     dataLayer.push(arguments);
   }
   gtag('js', new Date());
   gtag('config', 'G-XXXXXXXXXX');
 </script>
```

Link to README.md: https://github.com/amplitude/Amplitude-TypeScript/blob/AMP-81814/packages/plugin-ga-events-forwarder-browser/README.md

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  No

### CODE REVIEW

File: packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  /**
+   * Creates proxy for `navigator.sendBeacon` immediately to start listening for events.
+   * Google Analytics may start sending events before Amplitude SDK is initialized.
+   */
+  if (globalScope && isTrackingIdListValid) {
+    // eslint-disable-next-line @typescript-eslint/unbound-method
+    sendBeacon = globalScope.navigator.sendBeacon;
+
+    // eslint-disable-next-line @typescript-eslint/unbound-method
+    globalScope.navigator.sendBeacon = new Proxy(globalScope.navigator.sendBeacon, {
```
Comment (AlecBoutin) [inline packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts:54]: I feel like I saw GA send events via fetch/xhr at least once, did you happen to test for that?

**Review by AlecBoutin — APPROVED** [review]

**Review by justin-fiedler — COMMENTED** [review]
This is great @kevinpagtakhan!

File: packages/plugin-ga-events-forwarder-browser/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+## Usage
+
+This plugin works on top of Amplitude Browser SDK which listens for events tracked using Google Analytics and sends these events to Amplitude. To use this plugin, you need to install `@amplitude/analytics-browser` version `v1.9.1` or later.
+
+### 1. Import Amplitude packages
+
+* `@amplitude/plugin-ga-events-forwarder-browser`
+
+```typescript
+import { gaEventsForwarder } from '@amplitude/plugin-ga-events-forwarder-browser';
```
Comment (yuhao900914) [inline packages/plugin-ga-events-forwarder-browser/README.md:33]: Should this be gaEventsForwarderPlugin?

File: packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  trackingIds?: string | string[];
+}
+
+/**
+ * Returns an instance of `gaEventsForwarderPlugin`. Add this plugin to listen for events sent to Google Analytics,
+ * transform the events and send the events to Amplitude.
+ *
+ * ```html
+ * <script>
+ *   amplitude.add(gaEventsForwarder.plugin());
```
Comment (yuhao900914) [inline packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts:17]: is aEventsForwarder.plugin() right?

File: packages/plugin-ga-events-forwarder-browser/README.md (hunk 2 above)
Comment (kevinpagtakhan) [reply packages/plugin-ga-events-forwarder-browser/README.md:33]: you're right. i will update

File: packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts (hunk 3 above)
Comment (kevinpagtakhan) [reply packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts:17]: yes, this is how it looks like for script installation

File: packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+ *   gtag('config', 'G-XXXXXXXXXX');
+ * </script>
+ * ```
+ *
+ * @param options An object containing plugin options. Options include `trackingIds`.
+ *
+ * @param options.trackingIds A Google Analytics tracking ID or a list of Google Analytics tracking IDs.
+ * This limits the plugin to only listen for events tracked with the specified tracking ID/s.
+ */
+export const gaEventsForwarderPlugin = ({ trackingIds = [] }: Options = {}): BrowserEnrichmentPlugin => {
```
[the remark this answers is not in the document]
Comment (kevinpagtakhan) [reply packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts:36]: not really sure. it's neither an enrichment nor a destination plugin. they differ in how "execute" is defined, but this plugin will only have a passthrough (if enrichment) and noop (if destination). it uses the amplitude instance to send events, and would already use the amplitude destination plugin to send the events to amplitude.

File: packages/plugin-ga-events-forwarder-browser/package.json [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    "build:es5": "tsc -p ./tsconfig.es5.json",
+    "build:esm": "tsc -p ./tsconfig.esm.json",
+    "clean": "rimraf node_modules lib coverage",
+    "fix": "yarn fix:eslint & yarn fix:prettier",
+    "fix:eslint": "eslint '{src,test}/**/*.ts' --fix",
+    "fix:prettier": "prettier --write \"{src,test}/**/*.ts\"",
+    "lint": "yarn lint:eslint & yarn lint:prettier",
+    "lint:eslint": "eslint '{src,test}/**/*.ts'",
+    "lint:prettier": "prettier --check \"{src,test}/**/*.ts\"",
+    "test": "jest",
```
[the remark this answers is not in the document]
Comment (kevinpagtakhan) [reply packages/plugin-ga-events-forwarder-browser/package.json:32]: ah yes need to add that

File: packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts (hunk 1 above)
Comment (kevinpagtakhan) [reply packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts:54]: confirmed that it's for UA and not GA4

File: packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+   */
+  if (globalScope && isTrackingIdListValid) {
+    // eslint-disable-next-line @typescript-eslint/unbound-method
+    sendBeacon = globalScope.navigator.sendBeacon;
+
+    // eslint-disable-next-line @typescript-eslint/unbound-method
+    globalScope.navigator.sendBeacon = new Proxy(globalScope.navigator.sendBeacon, {
+      apply: (target, thisArg, [url, body]: [string, BodyInit | undefined]) => {
+        // Intercepts request and attempt to send to Amplitude
+        interceptRequest(new URL(url), body);
```
[the remark this answers is not in the document]
Comment (kevinpagtakhan) [reply packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts:57]: good point, i will update!

File: packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  /**
+   * 1. Parses the event payload of requests to GA4
+   * 2. Transforms GA4 events to Amplitude Events
+   * 3a: Pushes to preSetupEventQueue while waiting for Amplitude SDK to initialize
+   * 3b. Sends events to Amplitude after Amplitude SDK is initialized
+   */
+  const interceptRequest = (requestUrl: string, data?: BodyInit) => {
+    try {
```
Comment (kevinpagtakhan) [inline packages/plugin-ga-events-forwarder-browser/src/ga-events-forwarder.ts:71]: @AlecBoutin i added a try/catch here

**Review by justin-fiedler — APPROVED** [review]

**Review by AlecBoutin — APPROVED** [review]

**Pull request merged**